py/dml/types.py

```diff
@@ -88,6 +89,32 @@ def __str__(self):
 # anonymous structs used by global typedefs, label -> type
 global_anonymous_structs = {}
 
+def check_named_types(t):
+    '''Checks that a type does not reference a non-existing type'''
+    if isinstance(t, TNamed):
+        if t.c not in typedefs:
+            raise ETYPE(t.declaration_site, t)
+    elif isinstance(t, TStruct):
+        t.resolve()
+        for (mn, mt) in t.members.items():
+            check_named_types(mt)
+    elif isinstance(t, (TPtr, TVector, TArray)):
+        check_named_types(t.base)
+    elif isinstance(t, TFunction):
+        for pt in t.input_types:
+            check_named_types(pt)
+        check_named_types(t.output_type)
+    elif isinstance(t, TTraitList):
+        if t.traitname not in dml.globals.traits:
```

Comment:
perhaps an early assert that globals.traits is initialized?

Reply:
Considering `dml.globals.traits` starts off as `None` It would be noticeable either way (`t not in None` is not legal.) An assert does solidify the precondition, though. I'm ambivalent, I'll add an assert if you insist.

Reply:
then an explicit assert is not warranted. i think my only concern is that too early calls to this function are only captured if trait types are involved, making that hypothetical bug harder to catch. not a  disaster.

perhaps we should ideally have well-defined compile phases, and say that this function is only available after a particular phase.

Reply:
I very much like that idea. In particular, I've internally always lamented the fact that it's very unclear when `realtype` becomes safe to use. Having a phase after which you are guaranteed the ability to validate and normalize any type reference, and having `realtype` ICE if used before then, is a very tempting idea to me.

This'll mean rigorously splitting up the `structure` phase into subphases. What's nice I don't even think it'd take much for us to be able to establish these subphases: sure, there are some particular questions that complicate matters (e.g. there are some corners in resolving types besides `TNamed`, e.g. `Layout.resolve()` and `Hook.validate()`) but I think these are solvable. I'm thinking something like...
- Collecting all global declarations (in particular, templates and types)
- Resolve and verify global types
- Resolve and verify all other global declarations

... after which you deal with the device tree (which could also potentially be split into subphases)

Reply:
unfortunately, conditional imports make matters more complicated. so i guess it is not a topic for this PR.